src/strategies/base.ts

```diff
@@ -411,16 +410,22 @@ export abstract class BaseStrategy implements Strategy {
     return Version.parse('1.0.0');
   }
 
+  /**
+   * Adds a given file path to the strategy path.
+   * @param {string} file Desired file path.
+   * @returns {string} The file relative to the strategy.
+   */
   protected addPath(file: string) {
-    if (this.path === ROOT_PROJECT_PATH) {
-      return file;
+    // There is no strategy path to join, the strategy is at the root, or the
+    // file is at the root (denoted by a leading slash or tilde)
+    if (!this.path || this.path === ROOT_PROJECT_PATH || /^~?\//.test(file)) {
+      file = path.posix.normalize(file);
     }
-    file = file.replace(/^[/\\]/, '');
-    if (this.path === undefined) {
-      return file;
-    } else {
-      const path = this.path.replace(/[/\\]$/, '');
-      return `${path}/${file}`;
+    // Otherwise, the file is relative to the strategy path
+    else {
+      file = path.posix.join(this.path, file);
     }
+    // Strip any leading pathings (i.e. ~/, ../, /) and any trailing slashes and return
+    return file.replace(/^(\.{1,2}|~|\/)*\//, '').replace(/\/+$/, '');
```

Comment:
At this point the `file` has been normalized (in both invocations of `path.posix.normalize` and `path.posix.join`, e.g.

* `foo/bar/../baz` -> `foo/baz`
* `foo/bar/../../baz` -> `baz`
* `foo/bar/../../../../baz` -> `../../baz`

We can then safely strip any leading `../`, `/`, or `~/`, for example `../../baz` -> `baz` which would be at the root of the git repository.

![image](https://user-images.githubusercontent.com/7577851/148186249-f4ac04f8-095a-47c5-a35d-40a10ad31c97.png)

Reply:
shouldn't such paths be skipped?

Reply:
We could log a warning message and skip these files, but I think we could **only do this for extra files** updates - anywhere else `Strategy#addPath` is used (which appears to be a lot) probably aren't ready to receive `null` or catch an exception.

Reply:
This was my main reasoning for just silently ignoring any leading pathing - at least on Linux,
```
[jonathan@Alpine /]$ cd ../../../..
[jonathan@Alpine /]$ pwd
/
[jonathan@Alpine /]$ cd ~/../../../../../../../.././././../../
[jonathan@Alpine /]$ pwd
/
```

Reply:
I feel like using the path module is introducing some extra complexity and cases we don't want to handle. Do we ever need to use relative paths like `../something`? Probably not.

We can probably simplify the logic to:
* if it starts with `/`, then drop the slash and take the rest as the path to the file from the repository root
* otherwise, add the base path of the component

Reply:
I do think relative pathing can be useful, and in its current state, this works really well - as shown by the test cases I wrote. It supports jumping to the repo root, as well as using `../`, and as a side-effect of using this `path` module, it much more secure as it normalizes paths like `foo/../baz` to `baz`, meaning `foo/../../../../../etc/hosts` won't end up at `/etc/hosts`

Reply:
We could maybe disallow `../` in paths (for security), and instead require any relative paths to be absolute from the project root, e.g. for component `pkgs/b` instead of `../b.txt` you'd have to write `/pkgs/b.txt`.

Reply:
Yes, let's disallow extra relativity. We should make sure that `/` actually is relative from the git root and that paths without the leading `/` are relative to the component's root.

Reply:
@chingor13 removed the use of the `path` module in dc1f3adc4291f5721ffc3c9d022b320cf6c89124 - it now throws an `Error` if the resulting file path includes relativity, i.e. `../`